### PR TITLE
Change when we check the sampler.

### PR DESCRIPTION
**Pull request by bogdandrutu**
Change when do we apply the sampling decision and simplify the PrioritySampler to not depend on the previous sampling decision.

This is the first part of solving the issue https://github.com/census-instrumentation/opencensus-java/issues/656

### CODE REVIEW

File: impl_core/src/test/java/io/opencensus/implcore/trace/SpanBuilderImplTest.java [hunk 1]
```diff
@@ -164,6 +162,95 @@ public void startRemoteSpan() {
...
+        SpanBuilderImpl.createWithParent(SPAN_NAME, null, spanBuilderOptions).startSpan();
+    assertThat(rootSpan.getContext().isValid()).isTrue();
+    assertThat(rootSpan.getContext().getTraceOptions().isSampled()).isTrue();
+  }
+
+  @Test
+  public void startRemoteChildSpan_WithSpecifiedSampler() {
+    Span rootSpan =
+        SpanBuilderImpl.createWithParent(SPAN_NAME, null, spanBuilderOptions)
+            .setSampler(Samplers.neverSample())
```
Comment (sebright) [inline impl_core/src/test/java/io/opencensus/implcore/trace/SpanBuilderImplTest.java]: Should the parent use `alwaysSample`, if you are testing that the child ignores the parent's sampling decision?

File: impl_core/src/test/java/io/opencensus/implcore/trace/SpanBuilderImplTest.java [hunk 2]
```diff
@@ -164,6 +162,95 @@ public void startRemoteSpan() {
...
+    assertThat(childSpan.getContext().isValid()).isTrue();
+    assertThat(childSpan.getContext().getTraceId()).isEqualTo(rootSpan.getContext().getTraceId());
+    assertThat(childSpan.getContext().getTraceOptions().isSampled()).isTrue();
+  }
+
+  @Test
+  public void startChildSpan_WithSpecifiedSampler() {
+    Span rootSpan =
+        SpanBuilderImpl.createWithParent(SPAN_NAME, null, spanBuilderOptions)
+            .setSampler(Samplers.neverSample())
```
Comment (sebright) [inline impl_core/src/test/java/io/opencensus/implcore/trace/SpanBuilderImplTest.java]: Same here

File: impl_core/src/main/java/io/opencensus/implcore/trace/SpanBuilderImpl.java [hunk 3]
```diff
@@ -101,6 +99,30 @@ private SpanImpl startSpanInternal(
     return span;
   }
 
+  private static boolean makeSamplingDecision(
+      @Nullable SpanContext parent,
+      @Nullable Boolean hasRemoteParent,
+      String name,
+      Sampler sampler,
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply impl_core/src/main/java/io/opencensus/implcore/trace/SpanBuilderImpl.java]: Done.

File: api/src/test/java/io/opencensus/trace/samplers/SamplersTest.java [hunk 4]
```diff
@@ -114,59 +116,54 @@ public void probabilitySampler_outOfRangeLowProbability() {
...
-    double proportionSampled = (double) count / numSamples;
+    double proportionSampled = (double) count / NUM_SAMPLE_TRYES;
     // Allow for a large amount of slop (+/- 10%) in number of sampled traces, to avoid flakiness.
     assertThat(proportionSampled < probability + 0.1 && proportionSampled > probability - 0.1)
         .isTrue();
   }
 
   @Test
-  public void probabilitySamper_SamplesWithProbability() {
+  public void probabilitySampler_SamplesForUnsampledParent() {
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply api/src/test/java/io/opencensus/trace/samplers/SamplersTest.java]: Yes. Combined them.

File: api/src/test/java/io/opencensus/trace/samplers/SamplersTest.java [hunk 5]
```diff
@@ -114,59 +116,54 @@ public void probabilitySampler_outOfRangeLowProbability() {
...
-                  Collections.<Span>emptyList()))
-          .isTrue();
-    }
-  }
-
-  private final void probabilitySampler_SamplesWithProbabilityForUnsampled(
-      Sampler sampler, double probability) {
-    final int numSamples = 1000; // Number of traces for which to generate sampling decisions.
+  // Applies the given sampler to NUM_SAMPLE_TRYES random traceId/spanId pairs.
+  private static void applyProbabilitySampler(
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply api/src/test/java/io/opencensus/trace/samplers/SamplersTest.java]: Done.

File: api/src/test/java/io/opencensus/trace/samplers/SamplersTest.java [hunk 6]
```diff
@@ -33,6 +33,8 @@
 /** Unit tests for {@link Samplers}. */
 @RunWith(JUnit4.class)
 public class SamplersTest {
+  private static final String SPAN_NAME = "MySpanName";
+  private static final int NUM_SAMPLE_TRYES = 1000;
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply api/src/test/java/io/opencensus/trace/samplers/SamplersTest.java]: Done.

File: impl_core/src/test/java/io/opencensus/implcore/trace/SpanBuilderImplTest.java (hunk 1 above)
Comment (bogdandrutu) [reply impl_core/src/test/java/io/opencensus/implcore/trace/SpanBuilderImplTest.java]: Done.

File: impl_core/src/test/java/io/opencensus/implcore/trace/SpanBuilderImplTest.java (hunk 2 above)
Comment (bogdandrutu) [reply impl_core/src/test/java/io/opencensus/implcore/trace/SpanBuilderImplTest.java]: Done.

**Comment by bogdandrutu** [comment]
@adriancole they can do that by using AlwaysSample/NeverSample samplers if they have an external mechanism to do sampling.

Also I am working on some documentation about how the library does sampling.

**Pull request merged**